src/crypto/qat/qcccrypto.cc

```diff
@@ -20,51 +28,108 @@ static std::ostream& _prefix(std::ostream* _dout)
 // -----------------------------------------------------------------------------
 
 /*
- * Poller thread & functions
-*/
+ * Callback function
+ */
+static void symDpCallback(CpaCySymDpOpData *pOpData,
+                        CpaStatus status,
+                        CpaBoolean verifyResult)
+{
+  if (nullptr != pOpData->pCallbackTag)
+  {
+    static_cast<QatCrypto*>(pOpData->pCallbackTag)->complete();
+  }
+}
+
 static std::mutex qcc_alloc_mutex;
 static std::mutex qcc_eng_mutex;
 static std::atomic<bool> init_called = { false };
 
-void* QccCrypto::crypt_thread(void *args) {
-  struct qcc_thread_args *thread_args = (struct qcc_thread_args *)args;
-  thread_args->qccinstance->do_crypt(thread_args);
-  return thread_args;
+#define NON_INSTANCE -1
+#define RETRY_MAX_NUM 100
+
+template <typename CompletionToken>
+auto QccCrypto::async_get_instance(CompletionToken&& token) {
+  using boost::asio::async_completion;
+  using Signature = void(int);
+  async_completion<CompletionToken, Signature> init(token);
+
+  auto ex = boost::asio::get_associated_executor(init.completion_handler);
+
+  boost::asio::post(my_context, [this, ex, handler = std::move(init.completion_handler)]()mutable{
+    auto handler1 = std::move(handler);
+    if (!open_instances.empty()) {
+      int avail_inst = open_instances.front();
+      open_instances.pop_front();
+      boost::asio::post(ex, std::bind(handler1, avail_inst));
+    } else if (instance_completions.size() < max_queue_size) {
```

Comment:
if we change this to `if (!instance_completions.full())` i think we could avoid storing both `max_requests` and `max_queue_size` as member variables

Reply:
> if we change this to `if (!instance_completions.full())` i think we could avoid storing both `max_requests` and `max_queue_size` as member variables

Only can be remove `max_queue_size`.  `max_requests` is also used to call `init()` again after the first `init()` fails.